x/oracle/keeper/native_token.go

```diff
@@ -0,0 +1,242 @@
+		valueStakerList = k.cdc.MustMarshal(&stakerList)
+		store.Set(keyStakerList, valueStakerList)
+		stakerInfo.StakerIndex = int64(len(stakerList.StakerAddrs) - 1)
+	}
+
+	if newBalance.Balance <= 0 {
+		store.Delete(key)
+	} else {
+		stakerInfo.BalanceList = append(stakerInfo.BalanceList, &newBalance)
+		bz := k.cdc.MustMarshal(stakerInfo)
+		store.Set(key, bz)
+	}
+	return amount
+}
+
+// GetstakerList return stakerList for native-restaking asset of assetID
+func (k Keeper) GetStakerList(ctx sdk.Context, assetID string) types.StakerList {
+	store := ctx.KVStore(k.storeKey)
+	value := store.Get(types.NativeTokenStakerListKey(assetID))
+	if value == nil {
+		return types.StakerList{}
+	}
+	stakerList := &types.StakerList{}
+	k.cdc.MustUnmarshal(value, stakerList)
+	return *stakerList
+}
+
+// UpdateNativeTokenByBalanceChange updates balance info for staker under native-restaking asset of assetID when its balance changed by slash/refund on the source chain (beacon chain for eth)
+func (k Keeper) UpdateNativeTokenByBalanceChange(ctx sdk.Context, assetID string, rawData []byte, roundID uint64) error {
+	if len(rawData) < 32 {
+		return errors.New("length of indicate maps for stakers shoule be exactly 32 bytes")
+	}
+	sl := k.getStakerList(ctx, assetID)
+	if len(sl.StakerAddrs) == 0 {
+		return errors.New("staker list is empty")
+	}
+	stakerChanges, err := parseBalanceChange(rawData, sl)
+	if err != nil {
+		return err
+	}
+	store := ctx.KVStore(k.storeKey)
+	for stakerAddr, change := range stakerChanges {
+		key := types.NativeTokenStakerKey(assetID, stakerAddr)
+		value := store.Get(key)
+		if value == nil {
+			return errors.New("stakerInfo does not exist")
+		}
+		stakerInfo := &types.StakerInfo{}
+		k.cdc.MustUnmarshal(value, stakerInfo)
+		newBalance := types.BalanceInfo{}
+		if length := len(stakerInfo.BalanceList); length > 0 {
+			newBalance = *(stakerInfo.BalanceList[length-1])
+		}
+		newBalance.Block = uint64(ctx.BlockHeight())
+		if newBalance.RoundID == roundID {
+			newBalance.Index++
+		} else {
+			newBalance.RoundID = roundID
+			newBalance.Index = 0
+		}
+		newBalance.Change = types.BalanceInfo_ACTION_SLASH_REFUND
+		newBalance.Balance += int64(change)
+		stakerInfo.Append(&newBalance)
+		bz := k.cdc.MustMarshal(stakerInfo)
+		store.Set(key, bz)
+		// TODO: call assetsmodule. func(k Keeper) UpdateNativeRestakingBalance(ctx sdk.Context, stakerID, assetID string, amount sdkmath.Int) error
+	}
+	return nil
+}
+
+// getStakerList returns all Stakers for native-restaking of assetID, this is used for cache
+func (k Keeper) getStakerList(ctx sdk.Context, assetID string) types.StakerList {
+	if len(stakerList.StakerAddrs) == 0 {
+		stakerList = k.GetStakerList(ctx, assetID)
+	}
+	return stakerList
+}
+
+// parseBalanceChange parses rawData to details of amount change for all stakers relative to native restaking
+func parseBalanceChange(rawData []byte, sl types.StakerList) (map[string]int, error) {
+	indexs := rawData[:32]
```

Comment:
can you add some comment for this logic ?

Reply:
updated